### PR TITLE
Use ensemble evaluator in everest in place of BatchSimulator

### LINKED ISSUE #9308: Remove dependency on Simulator(BatchSimulator) and BatchContext

### PR DESCRIPTION
**Pull request by DanSava**
**Issue**
Resolves #9308 

**Approach**
_Short description of the approach_

(Screenshot of new behavior in GUI if applicable)

- [ ] PR title captures the intent of the changes, and is fitting for release notes.
- [ ] Added appropriate release note label
- [ ] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [ ] Make sure unit tests pass locally after every commit (`git rebase -i main
      --exec 'pytest tests/ert/unit_tests -n logical -m "not integration_test"'`)

## When applicable
- [ ] **When there are user facing changes**: Updated documentation
- [ ] **New behavior or changes to existing untested code**: Ensured that unit tests are added (See [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)).
- [ ] **Large PR**: Prepare changes in small commits for more convenient review
- [ ] **Bug fix**: Add regression test for the bug
- [ ] **Bug fix**: Create Backport PR to latest release

### CODE REVIEW

File: tests/everest/test_simulator_cache.py [hunk 1]
```diff
@@ -1,60 +1,67 @@
...
-                evaluator=simulator.create_forward_model_evaluator_function(),
-            )
-            .run()
-            .variables
-        )
-        assert variables2 is not None
-        assert n_evals == 0
-
-        assert np.array_equal(variables1, variables2)
+    # TODO reimplement test
```
Comment (yngve-sk) [inline tests/everest/test_simulator_cache.py:15]: What's the status on this file? Should it be deleted?

File: tests/everest/test_simulator_cache.py [hunk 2]
```diff
@@ -1,60 +1,67 @@
-import numpy as np
-from ropt.plan import BasicOptimizer
-
-from ert.storage import open_storage
-from everest.config import EverestConfig, SimulatorConfig
-from everest.optimizer.everest2ropt import everest2ropt
-from everest.simulator import Simulator
-from everest.simulator.everest_to_ert import everest_to_ert_config
+# import numpy as np
```
Comment (verveerpj) [inline tests/everest/test_simulator_cache.py:1]: This test needs to be fixed. If you need help, I can look into it.

File: src/everest/simulator/simulator.py [hunk 3]
```diff
@@ -1,300 +0,0 @@
-import datetime
```
Comment (verveerpj) [inline src/everest/simulator/simulator.py:1]: Empty file?

File: src/ert/run_models/everest_run_model.py [hunk 4]
```diff
@@ -131,160 +145,7 @@ def __call__(
```
Comment (verveerpj) [inline src/ert/run_models/everest_run_model.py:145]: Is the `MonitorThreadErrorCallback` protocol still needed?

File: tests/everest/test_simulator_cache.py (hunk 1 above)
Comment (verveerpj) [reply tests/everest/test_simulator_cache.py:15]: No, that feature should remain working, the tests needs to work.

File: src/ert/run_models/everest_run_model.py [hunk 5]
```diff
@@ -599,3 +463,306 @@ def result(self) -> Optional[OptimalResult]:
...
+                        )
+                    )
+                for var_name, var_setting in control.items():
+                    _check_suffix(ext_config, var_name, var_setting)
+
+                ensemble.save_parameters(
+                    control_name, sim_id, ExtParamConfig.to_dataset(control)
+                )
+
+    def create_forward_model_evaluator_function(self) -> Evaluator:
```
Comment (verveerpj) [inline src/ert/run_models/everest_run_model.py:596]: Why does this need to be a factory function? `run_forward_model` can just be a method. That will also make it easier to fix the `test_simulator_cache` test.

Comment (DanSava) [reply src/ert/run_models/everest_run_model.py:596]: There is no particular reason I just kept the old structure I thought it was useful for some tests if it can be made a simple method that would be fantastic.

File: src/ert/run_models/everest_run_model.py (hunk 4 above)
Comment (DanSava) [reply src/ert/run_models/everest_run_model.py:145]: No just leftovers I forgot to remove.

File: tests/everest/test_simulator_cache.py (hunk 1 above)
Comment (DanSava) [reply tests/everest/test_simulator_cache.py]: Peter fixed it

**Review by verveerpj — APPROVED** [review]
Good stuff!

File: tests/everest/test_data/open_shut_state_modifier/eclipse/model/INIT.SCH [hunk 6]
```diff
@@ -0,0 +1,23 @@
+WCONPROD
+ WELL-1 SHUT BHP 5* 100 /
+ WELL-2 SHUT BHP 5* 100 /
+ WELL-3 SHUT BHP 5* 100 /
+ WELL-4 SHUT BHP 5* 100 /
```
Comment (oyvindeide) [inline tests/everest/test_data/open_shut_state_modifier/eclipse/model/INIT.SCH:5]: Just a small comment, is this file needed?

**Review by oyvindeide — COMMENTED** [review]
Really nice work! Just had a tiny question

File: src/ert/run_models/base_run_model.py [hunk 7]
```diff
@@ -413,6 +413,12 @@ def _get_number_of_finished_realizations_from_reruns(self) -> int:
             False
         ) - self._initial_realizations_mask.count(False)
 
+    def get_current_snapshot(self) -> EnsembleSnapshot:
+        if self._iter_snapshot.keys():
+            current_iter = max(list(self._iter_snapshot.keys()))
+            return self._iter_snapshot[current_iter]
+        return EnsembleSnapshot()
```
Comment (oyvindeide) [inline src/ert/run_models/base_run_model.py:420]: Is this needed? This is for cases where the run model has no snapshots? Maybe it should be an error?

File: tests/everest/test_data/open_shut_state_modifier/eclipse/model/INIT.SCH (hunk 6 above)
Comment (DanSava) [reply tests/everest/test_data/open_shut_state_modifier/eclipse/model/INIT.SCH:5]: Yes it is needed for `test_state_modifier_workflow_run`. It was a case of two wrongs were making a right so the test was not testing anything because the expected resulting file was never created due to this file being removed.  

I added it back and also added a guard against this happening in the future for this particular test.

File: src/ert/run_models/base_run_model.py (hunk 7 above)
Comment (DanSava) [reply src/ert/run_models/base_run_model.py:420]: > Is this needed?

Not really, but makes handling calls easier. 

> This is for cases where the run model has no snapshots ?

No this is for cases when this is called before calling `run_experiment` on the run_model.

> Maybe it should be an error ?
Why? As far as I can tell there is no added benefit for throwing an error.
